### ISSUE TITLE
Blocking Entropy Side Channels

### ISSUE THREAD

**Issue by indolering** [WebAssembly/wasi-random#3]
In my brief literature review [0, 1, 2, 3, 4], one point of concern was that raw entropy and seeding events are side-channels.  This is a bit above my paygrade, but I believe that simply hashing the output from the raw entropy source should be sufficient to de-correlate both the actual bits from the raw entropy as well as the timing of **re**seeding events.

It's unclear to me how to hide the timing of the initialization phase during first boot. But as this is already a fairly minor side channel, I think that padding the timing would be a sufficient layer of tin-foil.

[0]: [SoK: Security Models for Pseudo-Random Number Generators](https://tosc.iacr.org/index.php/ToSC/article/view/606) ([PDF](https://tosc.iacr.org/index.php/ToSC/article/view/606/547))
[1]: [Recommendations for Randomness in the Operating System](https://www.usenix.org/conference/hotos15/workshop-program/presentation/corrigan-gibbs) ([PDF](https://people.csail.mit.edu/henrycg/files/academic/papers/hotos15recommendations.pdf))
[2]: [DJB, Entropy Attacks!](http://blog.cr.yp.to/20140205-entropy.html)
[3]: [Optimal Recovery Strategies for Compromised RNGs](https://link.springer.com/article/10.1007/s00453-016-0239-3) ([PDF](https://link.springer.com/content/pdf/10.1007/978-3-662-44381-1_3.pdf))
[4]: [Removing the Linux /dev/random blocking pool](https://lwn.net/Articles/808575/)

**Comment by indolering** [comment]
There may be complaints about not being able to perform entropy estimation.  This is not an unfounded concern, as whitening can prevent the detection of both [broken](https://arstechnica.com/gadgets/2019/10/how-a-months-old-amd-microcode-bug-destroyed-my-weekend/) and [backdoored](https://www.nytimes.com/2013/09/06/us/nsa-foils-much-internet-encryption.html) RNGs.  However, providing raw access to the entropy  would need to be modeled as access to the devices and processes generating the entropy.  Thus the discussion digresses to whether the host should be involved in entropy estimation at all.  Existential discussions on randomness are not very productive, as virtually all runtime environments provide an API for getting random bytes.

**Comment by sunfishcode** [comment]
Is this something that can be addressed at a spec level, or is this a quality-of-implementation concern?

**Comment by indolering** [comment]
I believe the spec should take side channels into account and even modeled formally.  The spec does not need to say _which_ hash function is used, but that output **must** be hashed.  The spec does not need to say how to ensure that the randomness provided to each process is uncorrelated, just that it **must** be uncorrelated.  I'm less certain on how to deal with timing analysis of calls to `rand`.

As discussed in chat, an elegant API for treating non-determinism and reproducibility as capabilities would be `random` and `unrandom` environmental variables containing seeds.  This would remove information about how long it takes to initialize the seed.

I don't think that the time it takes for the initial seeding to occur is a large side channel.  However, if `rand` is expected to return a fresh seed and blocks based on the entropy estimator, you open the door to building timers and compound problems with low quality RNGs.  You could block during the init phase, regularly reseed `rand`, and guarantee that the results are uncorrelated but **without** a guarantee that its returning a fresh seed.  

IMO, if a program like PGP **must** have an unused seed, then it should have to _ask_ for that capability.  I would want a formal security model to be able to account for the side channel's size, by rate limiting the call and padding return times.  The security masochist in me would want to make the capability configurable too 🤔.  If you deem a new, non-compatibility variant of `rand` necessary ... why not turn it into an async call and have it accept a caller provided salt (which can be ignored)?

**Comment by indolering** [comment]
What PGP and friends want is a `getEntropy` function that returns a fresh seed that is never reused and is not the output of a CSRNG ... which is not only a side channel but [a DoS vector](https://lwn.net/Articles/658497/).  However, they inevitably engage in entropy estimation on their own, [crawling up the stack and grabbing as many permissions as possible](https://lwn.net/Articles/808575/).  Instead providing a `getEntropy` API _designed_ for use in exceptional circumstances, maybe we should assume misuse and say "either assume the numbers coming off of our CSRNG are good and reseeding is happening optimally or roll your own".

Previously, I was thinking that `env.rand` (or `import sys.rand` or whatever) would be cut from the CSRNG, but we could specify that as providing a **single** use, fresh seed to each isolate as a middle ground.  This would still need to be modeled as a capability, as it still represents a side channel and DoS vector (albeit limited).  A best-effort `env.unrandom` seed cut from the CSRNG wouldn't need to block at all.  `rand()` would block until initialized, but then return values as fast as Keccak can spit them out.  Reseeding would be otherwise invisible to the client.

**Comment by sunfishcode** [comment]
Based on our conversations offline, I think it makes sense to say that TRNG / `getEntropy` APIs and userspace entropy estimation are things that WASI may eventually support, but which we can consider out of scope for the current iteration of the wasi-random proposal.

**Comment by indolering** [comment]
>  but which we can consider out of scope for the current iteration of the wasi-random proposal.

Agreed.  The spec should also forbid reseeding events from being exposed to clients.

**Comment by sunfishcode** [comment]
I agree that we should hide reseeding from clients.

I'm unfortunately not clear on what you mean by your second bullet point. My current belief is that if there's any viable workaround for insufficient entropy, it should be performed by the WASI implementation, and not by applications, as it's too complex to give applications adequate guidance on what kind of strategies they should employ if there is insufficient entropy.

**Comment by indolering** [comment]
> I'm unfortunately not clear on what you mean by your second bullet point. My current belief is that if there's any viable workaround for insufficient entropy, it should be performed by the WASI implementation, and not by applications, as it's too complex to give applications adequate guidance on what kind of strategies they should employ if there is insufficient entropy.

The second bullet point is just trying to reduce the bandwidth of any side channel that the timing of the initial seeding event would open up.

**Comment by sunfishcode** [comment]
Ah, ok. So the current approach is to just not have any asynchronous inits. The API is synchronous and infallible. It's expected to be backed by a seeded CSPRNG.

**Comment by indolering** [comment]
> Ah, ok. So the current approach is to just not have any asynchronous inits. The API is synchronous and infallible. It's expected to be backed by a seeded CSPRNG.

Right, so we agree on the following **mandatory** bits of the specification for `wasi-rand`:

* Synchronous API that **only** blocks until initially seeded.
  * ... i.e. it does **not** block _after_ initialization.
  * _Anyone wanting to emulate an async interface can just `import(Random.Rand).then(Rand => {...});`._
* Does not provide re-seeding information.
  * _... and return timing should not be impacted by the entropy collector or reseeding events in any way._
* Is thread safe.
  * _To prevent side channel attacks across threads._
* Is backed by a CSPRNG
* Whitens output / does **not** provide access to raw entropy / never supports any form of `getEntropy`.
  * _Presumably by hashing entropy via a cryptographically secure hashing algorithm._

Assuming that hashes are secure, the lack of access to reseeding events blocks all side channels **except** the timing of when `import(Random.Rand).then(Rand => {...});` returns after being blocked due to the initial seeding event.  Padding the initial return to the nearest even X milliseconds would reduce the size of the channel.

Since this is a small channel, no one else seems to care, and most implementations just want to pipe in the operating system's CSRNG it probably makes sense to add this as a _recommendation_ rather than a _requirement_.

If that sounds good to you, feel free to close the ticket.  If I ever get time, I'll try to update the README directly with the usual RFC must/should verbiage and add citations.

**Comment by sunfishcode** [comment]
> Synchronous API that only blocks until initially seeded.
>    ... i.e. it does not block after initialization.

The [currrent proposal](https://github.com/WebAssembly/wasi-random/blob/main/wasi-random.wit.md#getrandom) says that the function must not block, so if there's any initialization needed, it must be done before the application calls the API.

> Anyone wanting to emulate an async interface can just import(Random.Rand).then(Rand => {...});.

In Wasm, all the imports are resolved before an instance starts executing.

> Does not provide re-seeding information.
>    ... and return timing should not be impacted by the entropy collector or reseeding events in any way.
> Is thread safe.
>    To prevent side channel attacks across threads.
> Is backed by a CSPRNG
> Whitens output / does not provide access to raw entropy / never supports any form of getEntropy. 

The current text is [here](https://github.com/WebAssembly/wasi-random/blob/main/wasi-random.wit.md#getrandom). It does explicitly require a CSPRNG. And the API does not expose re-seeding or getEntropy.

It also requires the output to be "unpredictable", though that is vague and would benefit from clarification. It would be good to be clear that it doesn't repeat itself across threads, instances, etc.

And if there's a way to express the guidance that reseeding effects should leak as little timing information as possible in a brief sentence, then that'd be good to do too. Detailed discussions of timing are difficult to fit into a spec, and may be better suited for a standalone document about recommended implementation techniques.

**Comment by indolering** [comment]
> In Wasm, all the imports are resolved before an instance starts executing.

I assumed WASM loading mechanisms matched JavaScript's.  So there is no dynamic loading, everything *must* be declared upfront?

**Comment by sunfishcode** [comment]
That's right, at present there is no dynamic loading feature within WebAssembly.
